 **To predict the possibilty of a booking**

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import os
import scipy as sp
import warnings
import datetime
warnings.filterwarnings("ignore")
%matplotlib inline

In [2]:
from google.colab import drive                   #Connecting to google drive
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
ds1 = pd.read_csv("/content/drive/MyDrive/Data Science Projects/hotel_bookings.csv")    #Importing data set

In [6]:
ds1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

In [7]:
ds1.shape

(119390, 32)

In [10]:
ds1.isnull().all()

hotel                             False
is_canceled                       False
lead_time                         False
arrival_date_year                 False
arrival_date_month                False
arrival_date_week_number          False
arrival_date_day_of_month         False
stays_in_weekend_nights           False
stays_in_week_nights              False
adults                            False
children                          False
babies                            False
meal                              False
country                           False
market_segment                    False
distribution_channel              False
is_repeated_guest                 False
previous_cancellations            False
previous_bookings_not_canceled    False
reserved_room_type                False
assigned_room_type                False
booking_changes                   False
deposit_type                      False
agent                             False
company                           False


In [11]:
ds1=ds1.drop(['country','agent','company'],axis=1)

In [12]:
ds1=ds1.dropna()

In [13]:
ds1.isnull().sum()

hotel                             0
is_canceled                       0
lead_time                         0
arrival_date_year                 0
arrival_date_month                0
arrival_date_week_number          0
arrival_date_day_of_month         0
stays_in_weekend_nights           0
stays_in_week_nights              0
adults                            0
children                          0
babies                            0
meal                              0
market_segment                    0
distribution_channel              0
is_repeated_guest                 0
previous_cancellations            0
previous_bookings_not_canceled    0
reserved_room_type                0
assigned_room_type                0
booking_changes                   0
deposit_type                      0
days_in_waiting_list              0
customer_type                     0
adr                               0
required_car_parking_spaces       0
total_of_special_requests         0
reservation_status          

In [15]:
list_1=list(ds1.columns)     #lets find the categorial features

In [16]:
list_cate=[]
for i in list_1:
    if ds1[i].dtype=='object':
        list_cate.append(i)

In [17]:
from sklearn.preprocessing import LabelEncoder
le=LabelEncoder()

In [18]:
for i in list_cate:
    ds1[i]=le.fit_transform(ds1[i])

In [19]:
ds1

hotel  is_canceled  lead_time  arrival_date_year  arrival_date_month  \
0           1            0        342               2015                   5   
1           1            0        737               2015                   5   
2           1            0          7               2015                   5   
3           1            0         13               2015                   5   
4           1            0         14               2015                   5   
...       ...          ...        ...                ...                 ...   
119385      0            0         23               2017                   1   
119386      0            0        102               2017                   1   
119387      0            0         34               2017                   1   
119388      0            0        109               2017                   1   
119389      0            0        205               2017                   1   

        arrival_date_week_number  arrival_date_day_of_month  \
0                             27                          1   
1                             27                          1   
2                             27                          1   
3                             27                          1   
4                             27                          1   
...                          ...                        ...   
119385                        35                         30   
119386                        35                         31   
119387                        35                         31   
119388                        35                         31   
119389                        35                         29   

        stays_in_weekend_nights  stays_in_week_nights  adults  ...  \
0                             0                     0       2  ...   
1                             0                     0       2  ...   
2                             0                     1       1  ...   
3                             0                     1       1  ...   
4                             0                     2       2  ...   
...                         ...                   ...     ...  ...   
119385                        2                     5       2  ...   
119386                        2                     5       3  ...   
119387                        2                     5       2  ...   
119388                        2                     5       2  ...   
119389                        2                     7       2  ...   

        assigned_room_type  booking_changes  deposit_type  \
0                        2                3             0   
1                        2                4             0   
2                        2                0             0   
3                        0                0             0   
4                        0                0             0   
...                    ...              ...           ...   
119385                   0                0             0   
119386                   4                0             0   
119387                   3                0             0   
119388                   0                0             0   
119389                   0                0             0   

        days_in_waiting_list  customer_type     adr  \
0                          0              2    0.00   
1                          0              2    0.00   
2                          0              2   75.00   
3                          0              2   75.00   
4                          0              2   98.00   
...                      ...            ...     ...   
119385                     0              2   96.14   
119386                     0              2  225.43   
119387                     0              2  157.71   
119388                     0              2  104.40   
119389                     0              2  151.20   

        required_car_parking_spaces  total_of_special_r

In [20]:
y=ds1['hotel']
x=ds1.drop('hotel',axis=1)

**TRAINING AND TESTING DATA**

In [21]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,random_state=0,test_size=0.2)

In [22]:
print(len(x_train))
print(len(x_test))
print(len(y_train))
print(len(y_test))

95508
23878
95508
23878


# **MODELS**


## **1. Logistic Regression**

In [23]:
from sklearn.linear_model import LogisticRegression
reg = LogisticRegression()
reg.fit(x_train,y_train) 

LogisticRegression()

In [24]:
from sklearn.metrics import accuracy_score,confusion_matrix
from sklearn.metrics import classification_report
y_pred_reg=reg.predict(x_test)
acc_reg = accuracy_score(y_test, y_pred_reg)
print("Classification Report is:\n",classification_report(y_test,y_pred_reg))
print("Confusion Matrix:\n",confusion_matrix(y_test,y_pred_reg))
print("Training Score:\n",reg.score(x_train,y_train)*100)
print(f"Accuracy Score of Logistic Regression is : {acc_reg}")

Classification Report is:
               precision    recall  f1-score   support

           0       0.77      0.90      0.83     15891
           1       0.70      0.45      0.55      7987

    accuracy                           0.75     23878
   macro avg       0.73      0.68      0.69     23878
weighted avg       0.74      0.75      0.74     23878

Confusion Matrix:
 [[14375  1516]
 [ 4404  3583]]
Training Score:
 75.50362273317418
Accuracy Score of Logistic Regression is : 0.7520730379428763


## **2. KNeighborsClassifier**

In [25]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier()
knn.fit(x_train, y_train)

y_pred_knn = knn.predict(x_test)

acc_knn = accuracy_score(y_test, y_pred_knn)
conf = confusion_matrix(y_test, y_pred_knn)
clf_report = classification_report(y_test, y_pred_knn)

print(f"Accuracy Score of KNN is : {acc_knn}")
print(f"Confusion Matrix : \n{conf}")
print(f"Classification Report : \n{clf_report}")

Accuracy Score of KNN is : 0.8818996565876539
Confusion Matrix : 
[[14897   994]
 [ 1826  6161]]
Classification Report : 
              precision    recall  f1-score   support

           0       0.89      0.94      0.91     15891
           1       0.86      0.77      0.81      7987

    accuracy                           0.88     23878
   macro avg       0.88      0.85      0.86     23878
weighted avg       0.88      0.88      0.88     23878



## **3. DECISION TREE CLASSIFIER**

In [26]:
from sklearn.tree import DecisionTreeClassifier
dtree = DecisionTreeClassifier(max_depth=6, random_state=123,criterion='entropy')

dtree.fit(x_train,y_train)

DecisionTreeClassifier(criterion='entropy', max_depth=6, random_state=123)

In [27]:
y_pred=dtree.predict(x_test)
conf =print(confusion_matrix(y_test, y_pred))
clf =print(classification_report(y_test, y_pred))
score=accuracy_score(y_test,y_pred)
score

[[14919   972]
 [ 2448  5539]]
              precision    recall  f1-score   support

           0       0.86      0.94      0.90     15891
           1       0.85      0.69      0.76      7987

    accuracy                           0.86     23878
   macro avg       0.85      0.82      0.83     23878
weighted avg       0.86      0.86      0.85     23878



0.8567719239467292

## **4.Random Forest Classifier**

In [28]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV
rfc=RandomForestClassifier()
rfc.fit(x_train,y_train)

RandomForestClassifier()

In [29]:
RandomForestClassifier()

RandomForestClassifier()

In [30]:
y_pred=rfc.predict(x_test)
conf =print(confusion_matrix(y_test, y_pred))
clf =print(classification_report(y_test, y_pred))
score=accuracy_score(y_test,y_pred)
score

[[15570   321]
 [  507  7480]]
              precision    recall  f1-score   support

           0       0.97      0.98      0.97     15891
           1       0.96      0.94      0.95      7987

    accuracy                           0.97     23878
   macro avg       0.96      0.96      0.96     23878
weighted avg       0.97      0.97      0.97     23878



0.9653237289555239

In [31]:
ds1 = pd.DataFrame({'Actual': y_test, 'Predicted': y_pred})
ds1

Actual  Predicted
117018       0          0
107541       0          0
56907        0          0
26114        1          1
42024        0          0
...        ...        ...
55160        0          0
78598        0          0
98604        0          0
64854        0          0
51391        0          0

[23878 rows x 2 columns]

# **CONCLUSION**

So we get a accuracy score of 97 % using the Random Forest Classifier which is highest among all models